### ISSUE TITLE
KRM Exec Function can't locate executable when referencing a base

### ISSUE THREAD

**Issue by pmcjury** [kubernetes-sigs/kustomize#4347]
**Describe the bug**

When using a KRM exec function with a relative path to the _kustomization.yaml_ file, the executable can't be found when referencing a _base_ from a _variant_. 

It appears the function is "exec"-d relative to the base not the variant. 

Things are further complicated when the _base_ references another _remote base_. 

**Files that can reproduce the issue**

``` 

# dir
.
├── base
│   ├── deployment.yaml
│   └── kustomization.yaml
└── krm-exec
    ├── krm-config.yaml
    ├── kustomization.yaml
    └── plugins
        └── function.sh
```

```yaml

# base/kustomization.yaml

resources:
 - deployment.yaml

```

```yaml
# base/deployment.yaml

apiVersion: apps/v1
kind: Deployment
metadata:
  name: app1
spec:
  selector:
    matchLabels:
      app: hello
  template:
    metadata:
      labels:
        app: hello
    spec:
      containers:
      - name: busybox
        image: busybox:v1.0.0
```

```yaml

#krm-exec/kustomization.yaml
resources:
- ../base
transformers:
- krm-config.yaml
```

```yaml

#krm-exec/krm-config.yaml
apiVersion: my.example.io/v1
kind: MyPlugin
metadata:
  name: notImportantHere
  annotations:
    config.kubernetes.io/function: |
      exec: 
        path: ./plugins/function.sh
```

```sh
#krm-exec/plugins/function.sh
#!/bin/bash
resourceList=$(cat) 
# dummy
echo "$resourceList"
```

**Expected output**

The dummy function should just echo out the resources read from stdin. 

```yaml

apiVersion: apps/v1
kind: Deployment
metadata:
  name: app1
spec:
  selector:
    matchLabels:
      app: hello
  template:
    metadata:
      labels:
        any: thing
    spec:
      containers:
      - name: busybox
        image: busybox:v1.0.0
     
```

**Actual output**

```sh
#from krm-exec dir
kustomize build --enable-alpha-plugins --enable-exec  .
Error: couldn't execute function: fork/exec ./plugins/function.sh: no such file or directory

```

**Kustomize version**

```sh
kustomize version
{Version:kustomize/v4.4.1 GitCommit:b2d65ddc98e09187a8e38adc27c30bab078c1dbf BuildDate:2021-11-11T23:36:27Z GoOs:darwin GoArch:amd64}

```

**Platform**

macOS ( M1 )
darwin21.0
arm64

**Additional context**

**Some other scenarios:**

- The plugin works as expected when the variant kustomization does not reference the base, but a relative deployment.
```
#krm-exec/kustomizaiton.yaml
resources:
- deployment.yaml
```

- With the original setup If I move the `./plugins/function.sh` to the _base_, the plugin works even if the krm-function.yaml is located in the variant

To further complicate things, if I add a remote base to the local base I get this error:

```yaml
#base/kustomization.yaml
resources:
 - https://github.com/pmcjury/somerepo/manifests/base?ref=main
 - deployment.yaml
```

```
#from krm-exec dir
kustomize build --enable-alpha-plugins --enable-exec  .
Error: couldn't execute function: chdir /private/var/folders/yn/88ww9kqx1b35xylvc0wt0fb80000gn/T/kustomize-124974724/deploy/manifests/base: no such file or directory 
```

**Comment by pmcjury** [comment]
btw a work around is to just put the plugin in your $PATH and reference it without a path

**Comment by seh** [comment]
This looks like a duplicate of #4350. What timing!

**Comment by KnVerey** [comment]
/triage accepted

The fact that the Kustomization in question reading a child Kustomization interferes with the path is highly suspicous.

I think the root cause of both this and #4350 is that we're not properly deep copying the plugin loader here:

https://github.com/kubernetes-sigs/kustomize/blob/a8f58b4080748d9387cb54b98d7eb8b03f71046d/api/internal/target/kusttarget.go#L44-L45

Since the loader contains a pointer to a `PluginConfig`, duplicating it will still point to the same plugin config ([Simplified demo](https://go.dev/play/p/3jFSuNz3Ozs)). So when we reset the root in creating the child target, we end up modifying the plugin config for the parent. 😱 

@seh @pmcjury Since both issues are quite detailed (thank you!) I'm going to keep this one just because it was opened first. If either of you is interested in contributing a fix, please `/assign` yourself here. 

cc @natasha41575

**Comment by seh** [comment]
> a work around is to just put the plugin in your $PATH and reference it without a path

That's due to `exec.Cmd` using [`exec.LookPath`](https://pkg.go.dev/os/exec#LookPath) [here](https://cs.opensource.google/go/go/+/refs/tags/go1.17.5:src/os/exec/exec.go;l=174-177).

**Comment by seh** [comment]
@m-Bilal, have you had a chance to attempt a fix for this defect yet?

**Comment by aabouzaid** [comment]
If I get it right, all that's needed to fix this is just to change [the line no. 45](https://github.com/kubernetes-sigs/kustomize/blob/37a403fc65d417e591ee874d48081cac7986f513/api/internal/target/kusttarget.go#L47) from `pLdrCopy := *pLdr` to `pLdrCopy := pLdr`? :thinking: 

In that case, it will make a full copy instead of referring to the pointer.

If not, does anyone know what exactly is needed to fix this issue?

**Comment by seh** [comment]
> In that case, it will make a full copy instead of referring to the pointer.

It will make a _shallow_ copy. If you look at [the `loader.Loader` struct type](https://sourcegraph.com/github.com/kubernetes-sigs/kustomize@37a403fc65d417e591ee874d48081cac7986f513/-/blob/api/internal/plugins/loader/loader.go?L30-L38), you'll see that it's mostly pointers. Copying a `Loader` gives you the same pointer values, pointing at the same referenced values. Per Katrina's earlier https://github.com/kubernetes-sigs/kustomize/issues/4347#issuecomment-999960604, at minimum we need to copy the `types.PluginConfig` referenced by [the "pc" field](https://sourcegraph.com/github.com/kubernetes-sigs/kustomize@37a403fc65d417e591ee874d48081cac7986f513/-/blob/api/internal/plugins/loader/loader.go?L31).

Next, how we copy [a `types.PluginConfig`](https://sourcegraph.com/github.com/kubernetes-sigs/kustomize@37a403fc65d417e591ee874d48081cac7986f513/-/blob/api/types/pluginconfig.go?L12-L24)? Two of its fields are `int`s, two are structs. Both of the structs contain only simple scalar values, so copying a `types.PluginConfig` can be achieved via assignment.

I am not sure whether there are more fields that we'd need to copy deeply, but we do need to copy the `PluginConfig`, because its "FnLoadingOptions" field of type 'FnPluginLoadingOptions` has [a "WorkingDir" field](https://sourcegraph.com/github.com/kubernetes-sigs/kustomize@37a403fc65d417e591ee874d48081cac7986f513/-/blob/api/types/pluginrestrictions.go?L60-L61) that holds the crucial value, such that us using the wrong value there leads to this defect.

**Comment by aabouzaid** [comment]
@seh Thanks for your tip.
I've [fixed](https://github.com/kubernetes-sigs/kustomize/pull/4654) it by creating a new copy of `PluginConfig` with the values of `pc`.

I've tested it and it works fine now. If that approach is accepted, I will add the unit test for it.

**Comment by smoyer64** [comment]
This issue was killing me - I know functions are alpha but it took me a while to find this bug report.  Since I'm executing in a CD environment, changing the `PATH` has some unintended consequences.  An alternate work-around that avoids messing with the environment is to put all the plugins in one of the bases.  The caveat is that it has to be the last base that's loaded so I simply created a base that's nothing but `plugins/` and an empty `Kustomization` manifest.  My top-level resources look something like:

```
resources:
- ../base1/
- ../base2/
- ../plugins/
- manifest1.yaml
- manifest2.yaml
```

**Comment by aabouzaid** [comment]
@smoyer64 it's really annoying issue, and the fix is already there #4654
I hope to find someone to review the PR.

**Comment by annasong20** [comment]
/close

Issue should be fixed by #4654
